packages/devtools/lib/src/popup.dart

```diff
@@ -0,0 +1,632 @@
+
+    _poppedUp = false;
+  }
+
+  PopupListView get listView => _listView;
+}
+
+/// This class handles all the incremental matching as keys are types as well as
+/// navigation through the popup list e.g., pageUp, arrow up/down, etc.
+class AutoCompleteMatcher<T> {
+  AutoCompleteMatcher();
+
+  PopupListView _listView;
+  PopupListView get listView => _listView;
+
+  CoreElement _textfield; // Input element for keyboard input.
+  CoreElement get textfield => _textfield;
+
+  T _original;
+  int _originalScrollTop;
+
+  Map<String, List<T>> matchingState = {};
+
+  String _lastMatchingChars;
+
+  String get lastMatchingChars => _lastMatchingChars;
+
+  // Current Row via matching and navigation (up/down ARROW, up/down PAGE, HOME
+  // and END.
+  int _selectRow = -1;
+
+  StreamSubscription _subscription;
+
+  bool get active => _subscription != null;
+
+  FinishFunction _finishCallback;
+
+  void finish([bool cancel = false]) {
+    if (_finishCallback != null) _finishCallback(cancel);
+  }
+
+  void start(T revert, PopupListView<T> listView, CoreElement textfield,
+      [FinishFunction finishCallback]) {
+    _listView = listView;
+    _textfield = textfield;
+
+    if (finishCallback != null) _finishCallback = finishCallback;
+
+    _startMatching(revert, true);
+
+    // Start handling user's keystrokes to show matching list of files.
+    _subscription ??= _textfield.onKeyDown.listen((html.KeyboardEvent e) {
+      switch (e.keyCode) {
+        case DOM_VK_RETURN:
+          finish();
+          reset();
+          _listView.reset();
+          e.preventDefault();
+          break;
+        case DOM_VK_ESCAPE:
+          cancel();
+          break;
+        case DOM_VK_PAGE_UP:
+          _selectRow = _listView.page(ListDirection.pageUp, _selectRow);
+          e.preventDefault();
+          break;
+        case DOM_VK_PAGE_DOWN:
+          _selectRow = _listView.page(ListDirection.pageDown, _selectRow);
+          e.preventDefault();
+          break;
+        case DOM_VK_END:
+          _selectRow = _listView.page(ListDirection.end);
+          e.preventDefault();
+          break;
+        case DOM_VK_HOME:
+          _selectRow = _listView.page(ListDirection.home);
+          e.preventDefault();
+          break;
+        case DOM_VK_UP:
+          // Set selection one item up.
+          if (_selectRow > 0) {
+            _selectRow -= 1;
+            _listView.scrollAndHilight(_selectRow, -1);
+          }
+          e.preventDefault();
+          break;
+        case DOM_VK_DOWN:
+          // Set selection one item down.
+          if (_selectRow < listView.itemsAsList.length - 1) {
+            _selectRow += 1;
+            _listView.scrollAndHilight(_selectRow, -1);
+          }
+          e.preventDefault();
+          break;
+      }
+    });
+  }
+
+  void cancel() {
+    revert();
+    finish();
+  }
+
+  void selectFirstItem() {
+    _selectRow = 0;
+    _listView.scrollAndHilight(_selectRow, -1);
```

Comment:
is this supposed to be -1 or 1?

Reply:
Yes we just want to select item, scroll into view but not make the selected item become the first item.